Format mentions (without avatars) in the message list

danxuliu opened the pull request:
This pull request adds formatted mentions to the message list (but not formatted mentions nor auto-completion in messages being composed); it is based on #517, although in this case avatars are no longer included in the mention ([as decided in #517](https://github.com/nextcloud/spreed/pull/517#issuecomment-357277790)).

It is based on both the mentions shown in the Comments app and the rich formatting of notifications. The backend uses the Comments API to store the messages, so when `receiveMessages` is called the Comments API is used to extract the mentions from the plain text message originally sent. However, unlike the Comments app, the mentions are not returned to the client in its own attribute, but in a generic way using a rich object string. Thanks to this approach it should be easier to extend the formatting in the future with support for other [definition types](https://github.com/nextcloud/server/blob/0913bc2157293acafa186afec4018ad25d7fbb22/lib/public/RichObjectStrings/Definitions.php#L40).

In any case, as the mentions are got using the Comments API this system has its same limitations, like not supporting mentions to users with a space in the name.

Note that links are still formatted in the client using `OCP.Comments.plainToRich`, which formats the plain text links found in the message; this has not been modified, although for consistency it would be good to move the links too to the rich object string system (something for another pull request, though).

Before:
![chat-mention-no-avatar-before](https://user-images.githubusercontent.com/26858233/39135308-f013f990-4718-11e8-9804-9e80833d8aad.png)

After (outdated; now the display name is shown instead of the user name):
![chat-mention-no-avatar-after](https://user-images.githubusercontent.com/26858233/39135299-ea7b06cc-4718-11e8-9bc7-651c6c81b9c7.png)

Formatted mentions are just `span` elements with a `font-weight: bold` style; ~~they include the user name, but not the display name (it is the same as in GitHub, for example). Should the display name be shown too? Note that showing it in a tooltip may not be a good option, as (pending, to be done in another pull request) the contacts menu will be shown when clicking on the mention (like done in the Comments app), and the tooltip and the menu would look bad at the same time.~~ they include _@_ followed by the display name.

nickvergessen commented on js/richobjectstringparser.js:
```diff
@@ -0,0 +1,81 @@
+
+	OCA.SpreedMe.RichObjectStringParser = {
+
+		_userLocalTemplate: '<span class="mention-user">@{{id}}</span>',
```
We should use name here, so we display the user displayname, not the uid

danxuliu replied:
I think it is better to show **@nickvergesen** than **@Joas Schilling** or **Joas Schilling**, as the @ would make easier to spot mentions (I guess that in the future there could be other bold texts), and the @ prepended to the display name feels strange. Not a strong opinion on this, though; @nextcloud/designers ?

nickvergessen replied:
But `@some sha for ldap users` is not usable...

MorrisJobke replied:
It rather is `@597ae2f6-16a6-1027-98f4-d28b5365dc14` 😉

nickvergessen replied:
Since names are bold, I think it should be fine with an `@`.
I checked WhatsApp and Threema are showing you the "real"name, Telegram shows the id.
However just see the LDAP thing above and I guess everyone knows we shouldn't use IDs ;)

danxuliu replied:
>  Since names are bold, I think it should be fine with an @.

Maybe instead of just using bold text we could use [a coloured background like in Riot](https://github.com/nextcloud/server/pull/7514#issuecomment-356304123). But, for the time being, I think that just bold is fine ;-)

>  However just see the LDAP thing above and I guess everyone knows we shouldn't use IDs ;)

Agreed ;-) I will change it to use the real name.

danxuliu commented:
Oops, I will fix the tests.

nickvergessen commented:
One final thing:
https://github.com/nextcloud/spreed/blob/8821165b10b9f3ad466bd41604d93b4185de6617/js/views/chatview.js#L378-L384

This is not working anymore. I also failed to quickly move it over to the new code. Can you have a look? (It's the user menu which should be shown when you click on a mention)

danxuliu commented:
There are several issues with the contacts menu unrelated to the mentions themselves, so I intended to fix them and make the contacts menu appear when clicking on a mention in a different pull request.

nickvergessen commented on lib/Controller/ChatController.php:
```diff
@@ -247,14 +253,17 @@ public function receiveMessages($token, $lookIntoFuture, $limit = 100, $lastKnow
 				'timestamp' => $comment->getCreationDateTime()->getTimestamp(),
-				'message' => $comment->getMessage()
+				'message' => $message,
+				'messageParameters' => $messageParameters,
```
Should adjust the `docs/api-v1.md`

danxuliu replied:
Do you prefer me to explicitly document all the possible parameters or just point to the _RichMessageHelper.php_ file?

nickvergessen replied:
No, just that there is `messageParameters`
See rich subject of activity: https://github.com/nextcloud/activity/blob/master/docs/endpoint-v2.md#activity-element
Maybe you can also link to https://github.com/nextcloud/server/issues/1706

danxuliu replied:
Done; I shamelessly ripped it from [the Notifications app](https://github.com/nextcloud/notifications/commit/19cc41a9ceb4c68e5b3093c8bd6c04d6541a880a#diff-be2c9ef579bbe4a1f1b0009530d415faR121)

MorrisJobke commented:
> For now this seems good. We need to have a broader discussion regarding showing usernames vs real names (or both). Usually we show real names (if present) everywhere in sharing, and as name next to the avatar.

I don't think there is much to discuss here. Username is purely internally. All the user cares about is the login name (could be username, email or any other attribute in case of LDAP) and the display name. The username (our internal representation of a user) could be highly artificial - like `@597ae2f6-16a6-1027-98f4-d28b5365dc14` which is the default for every LDAP setup.

jancborchardt commented:
@MorrisJobke yeah, that was kinda my point. So we should not use the username here, but the real name, right? (However it needs to fall back to username if no real name is set.)

danxuliu commented:
I have fixed the failing acceptance tests and also added _rich-strings_ to the capabilities (as I supposed that it had to be added; if not just remove the last commit ;-) ).

Backport is ready in https://github.com/nextcloud/spreed/tree/stable13-805-format-mentions-without-avatars-in-the-message-list (no pull request until this one is merged, though).
